src/ring_theory/hahn_series.lean

```diff
@@ -0,0 +1,390 @@
+
+/-- `single a r` is the Hahn series which has coefficient `r` at `a` and zero otherwise. -/
+def single (a : Γ) (r : R) : hahn_series Γ R := {
+  coeff := λ b, if a = b then r else 0,
+  is_wf_support := (set.is_wf_singleton a).mono (λ b, begin
+    simp only [and_imp, exists_prop, set.mem_singleton_iff, ite_eq_right_iff,
+      ne.def, set.mem_set_of_eq, not_forall],
+    exact λ h1 _, h1.symm,
+  end) }
+
+variables {a b : Γ} {r : R}
+
+theorem single_coeff : (single a r).coeff b = if a = b then r else 0 := rfl
+
+@[simp]
+theorem single_coeff_same : (single a r).coeff a = r := if_pos rfl
+
+@[simp]
+theorem single_coeff_of_ne (h : a ≠ b) : (single a r).coeff b = 0 := if_neg h
+
+@[simp]
+lemma single_support_of_ne (h : r ≠ 0) : {b | (single a r).coeff b ≠ 0} = {a} :=
+begin
+  ext b,
+  simp only [set.mem_singleton_iff, ne.def, set.mem_set_of_eq],
+  split,
+  { contrapose!,
+    exact λ h', single_coeff_of_ne (ne.symm h') },
+  { rintro rfl,
+    simp [h] }
+end
+
+@[simp]
+lemma single_eq_zero : (single a (0 : R)) = 0 := by { ext, simp [single_coeff] }
+
+end zero
+section addition
+
+variable [linear_order Γ]
+
+instance [add_monoid R] : has_add (hahn_series Γ R) :=
+{ add := λ x y, { coeff := x.coeff + y.coeff,
+                  is_wf_support := (x.is_wf_support.union y.is_wf_support).mono (λ a ha, begin
+                    contrapose! ha,
+                    simp only [not_or_distrib, not_not, set.mem_union_eq, set.mem_set_of_eq] at ha,
+                    simp only [pi.add_apply, not_not, set.mem_set_of_eq],
+                    rw [ha.1, ha.2, add_zero],
+                  end) } }
+
+instance [add_monoid R] : add_monoid (hahn_series Γ R) :=
+{ zero := 0,
+  add := (+),
+  add_assoc := λ x y z, by { ext, apply add_assoc },
+  zero_add := λ x, by { ext, apply zero_add },
+  add_zero := λ x, by { ext, apply add_zero } }
+
+@[simp]
+lemma add_coeff [add_monoid R] {x y : hahn_series Γ R} {a : Γ} :
+  (x + y).coeff a = x.coeff a + y.coeff a := rfl
+
+instance [add_comm_monoid R] : add_comm_monoid (hahn_series Γ R) :=
+{ add_comm := λ x y, by { ext, apply add_comm }
+  .. hahn_series.add_monoid }
+
+instance [add_group R] : add_group (hahn_series Γ R) :=
+{ neg := λ x, { coeff := λ a, - x.coeff a,
+                is_wf_support := by { convert x.is_wf_support,
+                  ext a,
+                  simp }, },
+  add_left_neg := λ x, by { ext, apply add_left_neg },
+  .. hahn_series.add_monoid }
+
+@[simp]
+lemma neg_coeff [add_group R] {x : hahn_series Γ R} {a : Γ} : (- x).coeff a = - x.coeff a := rfl
+
+instance [add_comm_group R] : add_comm_group (hahn_series Γ R) :=
+{ .. hahn_series.add_comm_monoid,
+  .. hahn_series.add_group }
+
+end addition
+section smul
+
+variables [linear_order Γ] {V : Type*}
+
+section distrib_mul_action
+variables [monoid R] [add_monoid V] [distrib_mul_action R V]
+
+instance : has_scalar R (hahn_series Γ V) :=
+⟨λ r x, { coeff := λ a, r • x.coeff a,
```

Comment:
```suggestion
⟨λ r x, { coeff := r • x.coeff,
```

Reply:
This one doesn't compile.

Reply:
You'll need `import algebra.module.pi` I think.

Reply:
Thanks!